File: forUWP/KaruahChess/rules/BoardAnimation.cs
﻿/*
Karuah Chess is a chess playing program
Copyright (C) 2020 Karuah Software

Karuah Chess is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

Karuah Chess is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

using System;
using KaruahChess.Pieces;
using Windows.UI.Xaml.Media.Imaging;
using KaruahChess.Model;
using KaruahChessEngine;
using KaruahChess.Common;
using System.Collections.Generic;

namespace KaruahChess.Rules
{
    public class BoardAnimation
    {

        KaruahChessEngineClass _tempBoardBefore;
        KaruahChessEngineClass _tempBoardAfter;

        /// <summary>
        /// Constructor
        /// </summary>
        public BoardAnimation()
        {
            _tempBoardBefore = new KaruahChessEngineClass();
            _tempBoardAfter = new KaruahChessEngineClass();
        }

        /// <summary>
        /// Creates a move animation sequence
        /// </summary>
        /// <param name="pBoardA"></param>
        /// <param name="pBoardB"></param>
        /// <param name="pBoardSquareDS"></param>
        /// <returns></returns>
        public List<PieceAnimationInstruction> CreateAnimationList(GameRecordArray pBoardRecA, GameRecordArray pBoardRecB, BoardSquareDataService pBoardSquareDS)
        {
            var animationList = new List<PieceAnimationInstruction>(4);
            var moveList = GetAnimationMoveList(pBoardRecA, pBoardRecB);                        
            int animationNumber = 0;

            
            foreach (var move in moveList) {
                int spin = move[0];
                int fromIndex = move[1];
                int toIndex = move[2];

                if (fromIndex > -1 && toIndex > -1) {
                    // Piece move animation                                          
                    var pieceImgData = Piece.GetImage(BoardSquareDataService.GetPieceTypeFromSpin(spin), BoardSquareDataService.GetPieceColourFromSpin(spin), pBoardSquareDS.SquareSize, pBoardSquareDS.SquareSize);
                    var instruction = GetAnimationInstruction(fromIndex, toIndex, pieceImgData, pBoardSquareDS);
                    instruction.AnimationType = PieceAnimationInstruction.AnimationTypeEnum.Move;
                    animationList.Add(instruction);                    
                }
                else if (fromIndex > -1 && toIndex == -1)
                {
                    // Piece take animation
                    var pieceImgData = Piece.GetImage(BoardSquareDataService.GetPieceTypeFromSpin(spin), BoardSquareDataService.GetPieceColourFromSpin(spin), pBoardSquareDS.SquareSize, pBoardSquareDS.SquareSize);
                    var instruction = GetAnimationInstruction(fromIndex, fromIndex, pieceImgData, pBoardSquareDS);
                    instruction.AnimationType = PieceAnimationInstruction.AnimationTypeEnum.Take;
                    animationList.Add(instruction);
                }
                else if (fromIndex == -1 && toIndex > -1)
                {
                    // Piece return animation                    
                    var pieceImgData = Piece.GetImage(BoardSquareDataService.GetPieceTypeFromSpin(spin), BoardSquareDataService.GetPieceColourFromSpin(spin), pBoardSquareDS.SquareSize, pBoardSquareDS.SquareSize);
                    var instruction = GetAnimationInstruction(toIndex, toIndex, pieceImgData, pBoardSquareDS);
                    instruction.AnimationType = PieceAnimationInstruction.AnimationTypeEnum.Put;
                    animationList.Add(instruction);
                }
                animationNumber++;
            }
            
            return animationList;
        }
               

        /// <summary>
        /// Detects changes between two bitboards and returns list of moves. Only works up to two different piece types. If
        /// more than two pieces changed, then the list returned is empty.
        /// </summary>
        /// <param name="pBoardFrom"></param>
        /// <param name="pBoardTo"></param>
        /// <returns></returns>
        private List<int[]> GetAnimationMoveList(GameRecordArray pBoardRecBefore, GameRecordArray pBoardRecAfter)
        {
            _tempBoardBefore.SetBoardArray(pBoardRecBefore.BoardArray);
            _tempBoardBefore.SetStateArray(pBoardRecBefore.StateArray);
            _tempBoardAfter.SetBoardArray(pBoardRecAfter.BoardArray);
            _tempBoardAfter.SetStateArray(pBoardRecAfter.StateArray);

            UInt64 allBeforeWhitePos = _tempBoardBefore.GetOccupiedByWhite();
            UInt64 allBeforeBlackPos = _tempBoardBefore.GetOccupiedByBlack();
            UInt64 allAfterWhitePos = _tempBoardAfter.GetOccupiedByWhite();
            UInt64 allAfterBlackPos = _tempBoardAfter.GetOccupiedByBlack();

            UInt64 allChangeWhitePos = allBeforeWhitePos ^ allAfterWhitePos;
            UInt64 allChangeBlackPos = allBeforeBlackPos ^ allAfterBlackPos;

           
            int[,] spinChange = new int[13, 2];
            CreateChangeArray(allChangeWhitePos, spinChange);           
            CreateChangeArray(allChangeBlackPos, spinChange);
            List<int[]> spinChangeList = ConvertSpinChangeArrayToList(spinChange);
                        
            // Too many changes to animate so just clear the list
            if (spinChangeList.Count > 2)
            {
                spinChangeList.Clear();
            }

            return spinChangeList;
        }

        /// <summary>
        /// Converts a spin change array to a list of changes
        /// </summary>
        /// <param name="pSpinChange"></param>
        /// <returns></returns>
        private List<int[]> ConvertSpinChangeArrayToList(int[,] pSpinChange)
        {
            const int spinOffset = 6;
            const int sqIndexOffset = 1;

            // Build the list of changes
            List<int[]> pieceChangeList = new List<int[]>(2);
            for (int i = 0; i < 13; i++)
            {
                if (pSpinChange[i, 0] > 0 || pSpinChange[i, 1] > 0)
                {
                    int[] change = new int[] { i - spinOffset, -1, -1 };
                    if (pSpinChange[i, 0] > 0) change[1] = pSpinChange[i, 0] - sqIndexOffset;
                    if (pSpinChange[i, 1] > 0) change[2] = pSpinChange[i, 1] - sqIndexOffset;
                    pieceChangeList.Add(change);
                }
            }

            return pieceChangeList;
        }

        /// <summary>
        /// Creates an array of spin changes
        /// </summary>
        /// <param name="pChangedPositions"></param>
        /// <returns></returns>        
        private void CreateChangeArray(UInt64 pChangedPositions, int[,] pSpinChange)
        {
            // [Spin offset, {From Index, To Index}]
            
            const int spinOffset = 6;
            const int sqIndexOffset = 1;

            // Changes            
            while (pChangedPositions > 0)
            {  

                int pos = helper.BitScanForward(pChangedPositions);
                UInt64 sqMask = 1uL << pos;
                pChangedPositions ^= sqMask;
                int sqIndex = 63 - pos;

                int beforeSpin = _tempBoardBefore.GetSpin(sqIndex);
                int afterSpin = _tempBoardAfter.GetSpin(sqIndex);

                if (beforeSpin != 0) pSpinChange[beforeSpin + spinOffset, 0] = sqIndex + sqIndexOffset;
                if (afterSpin != 0) pSpinChange[afterSpin + spinOffset, 1] = sqIndex + sqIndexOffset;
                                
            }
                        
        }


        /// <summary>
        /// Creates an animation instruction
        /// </summary>
        /// <param name="pFromId"></param>
        /// <param name="pToId"></param>
        /// <param name="pImage"></param>
        /// <param name="pImageRotationZ"></param>
        /// <returns></returns>
        private PieceAnimationInstruction GetAnimationInstruction(int pFromIndex, int pToIndex, BitmapImage pImage, BoardSquareDataService pBoardSquareDS)
        {
            var fromTile = pBoardSquareDS.BoardTiles[pFromIndex];
            var toTile = pBoardSquareDS.BoardTiles[pToIndex];
            var fromTilePoint = fromTile.Coordinates();
            var toTilePoint = toTile.Coordinates();

            var animIntruct = new PieceAnimationInstruction();
            animIntruct.ImageData = pImage;
            animIntruct.MoveFrom = fromTilePoint;
            animIntruct.MoveTo = toTilePoint;
            animIntruct.HiddenSquareIndexes.Add(pFromIndex);
            animIntruct.HiddenSquareIndexes.Add(pToIndex);

            return animIntruct;
        }
    }
}

File: forAndroid/app/src/main/java/purpletreesoftware/karuahchess/rules/BoardAnimation.kt
/*
Karuah Chess is a chess playing program
Copyright (C) 2020 Karuah Software

Karuah Chess is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

Karuah Chess is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

package purpletreesoftware.karuahchess.rules

import android.content.Context
import android.graphics.drawable.Drawable
import androidx.core.content.ContextCompat
import purpletreesoftware.karuahchess.engine.KaruahChessEngineC
import purpletreesoftware.karuahchess.model.gamerecord.GameRecordArray
import purpletreesoftware.karuahchess.customcontrol.TileAnimationInstruction
import purpletreesoftware.karuahchess.customcontrol.TileAnimationSequence
import purpletreesoftware.karuahchess.customcontrol.TilePanel

@ExperimentalUnsignedTypes
class BoardAnimation {

    companion object {

        val _tempBoardFrom: KaruahChessEngineC = KaruahChessEngineC()
        val _tempBoardTo: KaruahChessEngineC = KaruahChessEngineC()
        /**
         * Creates a move animation sequence
         */
        fun createAnimationMoveSequence(pBoardRecA : GameRecordArray, pBoardRecB : GameRecordArray, pTilePanel: TilePanel, pContext: Context) : TileAnimationSequence
        {
            val moveDict = getAnimationMoveList(pBoardRecA, pBoardRecB)
            var animationNumber = 0
            val animationSequence: TileAnimationSequence = TileAnimationSequence()

            for (move in moveDict)
            {
                val fromIndex = move.value[0]
                val toIndex = move.value[1]
                val fromSpin = move.value[2]
                val toSpin = move.key

                if (fromIndex > -1 && toIndex >-1 && fromSpin != 0)
                {
                    // Piece move animation
                    val pieceImage = ContextCompat.getDrawable(pContext, pTilePanel.getImage(fromSpin))
                    if (pieceImage != null) {
                        val animationInstruction =
                            getAnimationInstruction(fromIndex, toIndex, pTilePanel, pieceImage, pContext)
                        if (animationInstruction != null) {
                            if (animationNumber == 0) animationSequence.add(
                                TileAnimationSequence.AnimationSeqEnum.A,
                                animationInstruction
                            )
                            else if (animationNumber == 1) animationSequence.add(
                                TileAnimationSequence.AnimationSeqEnum.B,
                                animationInstruction
                            )
                        }
                        animationNumber++
                        if (animationNumber > 1) break
                    }
                }
                else if (fromIndex > -1 && toIndex == -1 && toSpin != 0)
                {
                    // Piece take animation
                    val pieceImage = ContextCompat.getDrawable(pContext, pTilePanel.getImage(fromSpin))
                    if (pieceImage != null) {
                        val animationInstruction = getAnimationInstruction(fromIndex, fromIndex, pTilePanel, pieceImage, pContext)
                        if (animationInstruction != null) {
                            animationSequence.add(TileAnimationSequence.AnimationSeqEnum.C, animationInstruction)
                        }
                    }
                }
                else if (fromIndex == -1 && toIndex > -1 && toSpin != 0)
                {
                    // Piece return animation
                    val pieceImage = ContextCompat.getDrawable(pContext, pTilePanel.getImage(toSpin))
                    if (pieceImage != null) {
                        val animationInstruction = getAnimationInstruction(toIndex, toIndex, pTilePanel, pieceImage, pContext)
                        if (animationInstruction != null) {
                            animationSequence.add(TileAnimationSequence.AnimationSeqEnum.D, animationInstruction)
                        }
                    }
                }

            }

            return animationSequence
        }

        /**
         * Create fall animation
         */
        fun createAnimationFall(pIndex: Int, pTilePanel: TilePanel, pContext: Context) : TileAnimationSequence
        {

            val animationSequence = TileAnimationSequence()
            val spin = pTilePanel.getTile(pIndex)?.spin

            if (pIndex > -1 && spin != null && spin != 0)
            {
                // fall animation
                val pieceImage = ContextCompat.getDrawable(pContext, pTilePanel.getImage(spin))
                if (pieceImage != null) {
                    val animationInstruction = getAnimationInstruction(pIndex, pIndex, pTilePanel, pieceImage, pContext)
                    if (animationInstruction != null) {
                        animationSequence.add(TileAnimationSequence.AnimationSeqEnum.E, animationInstruction)
                    }
                }
            }

            return animationSequence
        }

        /**
         * Creates an animation instruction
         */
        private fun getAnimationInstruction(
            pFromIndex: Int,
            pToIndex: Int,
            pTilePanel: TilePanel,
            pPieceImage: Drawable,
            pContext: Context
        ): TileAnimationInstruction? {
            if (pFromIndex > -1 && pFromIndex < 64 && pToIndex > -1 && pToIndex < 64) {
                val fromTile = pTilePanel.getTile(pFromIndex)
                val toTile = pTilePanel.getTile(pToIndex)

                if (fromTile?.piece != null && toTile?.piece != null) {
                    val animInstruct = TileAnimationInstruction(pContext)
                    animInstruct.imageData.setImageDrawable(pPieceImage)
                    animInstruct.moveFromX = fromTile.x
                    animInstruct.moveFromY = fromTile.y
                    animInstruct.moveToX = toTile.x
                    animInstruct.moveToY = toTile.y

                    animInstruct.hiddenSquareIndexes.add(pFromIndex)
                    animInstruct.hiddenSquareIndexes.add(pToIndex)

                    return animInstruct
                }

            }

            return null
        }

        /**
         * Detects changes between two bitboards and returns list of moves. Only works up to two different pieces types. If
         *  more than two pieces changed, then the list returned is empty.
         */
        fun getAnimationMoveList(pBoardRecFrom: GameRecordArray, pBoardRecTo: GameRecordArray): HashMap<Int, IntArray>
        {


            _tempBoardFrom.setBoardArray(pBoardRecFrom.boardArray)
            _tempBoardFrom.setStateArray(pBoardRecFrom.stateArray)

            _tempBoardTo.setBoardArray(pBoardRecTo.boardArray)
            _tempBoardTo.setStateArray(pBoardRecTo.stateArray)


            val moveDict = HashMap<Int, IntArray>(2)
            var tooManyChanges = false

            //Loop through bit boards and detectchanges

            // Find from index move
            for (i in 0..63)
            {
                val fromSpin = _tempBoardFrom.getSpin(i)
                val toSpin = _tempBoardTo.getSpin(i)
                if (toSpin != fromSpin && fromSpin != 0)
                {
                    if (moveDict.containsKey(fromSpin))
                    {
                        tooManyChanges = true
                        moveDict.clear()
                        break
                    }
                    else
                    {
                        val move = intArrayOf( -1, -1, 0 ) //from, to, from spin, to spin
                        move[0] = i
                        move[2] = fromSpin
                        moveDict[fromSpin] = move
                    }
                }
            }

            // If too many changes just clear and return empty dictionary
            if (tooManyChanges)
            {
                moveDict.clear()
                return moveDict
            }

            // Find to index move
            for (i in 0..63)
            {
                val fromSpin = _tempBoardFrom.getSpin(i)
                val toSpin = _tempBoardTo.getSpin(i)
                if (fromSpin != toSpin && toSpin != 0)
                {
                    if (moveDict.containsKey(toSpin))
                    {
                        val move = moveDict[toSpin]
                        if (move != null) move[1] = i
                    }
                    else
                    {
                        val move = IntArray(3)
                        move[0] = -1
                        move[1] = i
                        move[2] = toSpin
                        moveDict[toSpin] = move
                    }
                }
            }

            // Limit to two changes
            if (moveDict.count() > 2) moveDict.clear()



            return moveDict
        }




    }

}